### PR TITLE
Upload files to Amazon S3 support

### PR DESCRIPTION
**Pull request by sampaiodiego**
- [x] Support S3 config
- [x] Protect access to files uploaded to S3
- [x] CDN configuration

image

### CODE REVIEW

**Comment by graywolf336** [comment]
:+1: Looking forward to this!

**Comment by engelgabriel** [comment]
@RocketChat/core please review?

**Comment by geekgonecrazy** [comment]
LGTM

@sampaiodiego you rock!  This is an amazing change!

**Comment by Sing-Li** [comment]
This is TOTALLY awesome!    One request, can you please make the `bucketURL` and `region` from Slingshot configurable in settings?   They are essential for configuring lower-cost S3 'compatible' providers - although they'd always be the same for AWS.

File: packages/rocketchat-file-upload/server/lib/responseS3.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+});
+RocketChat.settings.get('FileUpload_S3_AWSAccessKeyId', function(key, value) {
+	S3accessKey = value;
+});
+RocketChat.settings.get('FileUpload_S3_AWSSecretAccessKey', function(key, value) {
+	S3secretKey = value;
+});
+
+var generateURL = function(fullUrl) {
+	var resourceURL = '/' + fullUrl.substr(fullUrl.indexOf(S3bucket));
```
Comment (nikashitsa) [inline packages/rocketchat-file-upload/server/lib/responseS3.js]: resourceURL gives string like that for me `/bucket-name.s3.amazonaws.com/ZkchB8J9Nk3neCEHy/HcY5AuFM66M5hkfX6/esH5Y2FiiWvav8yBW`
which gives wrong signature and 403 error.
resourceURL should be like that `/bucket-name/ZkchB8J9Nk3neCEHy/HcY5AuFM66M5hkfX6/esH5Y2FiiWvav8yBW`

Possible fix:
`var resourceURL = '/' + S3bucket + fullUrl.substr(fullUrl.indexOf('s3.amazonaws.com')+16);`

@sampaiodiego please have a look.

**Comment by sampaiodiego** [comment]
@nikashitsa thanks for the input.

please, can you test with the new code I've just pushed?

**Comment by nikashitsa** [comment]
@sampaiodiego my error has gone. Thank you.
But after every page refresh I'm getting to login page. It's ok?

**Comment by sampaiodiego** [comment]
this bug was introduced by a8423219f062e4f1c64f559e59cb5cf09b0d2272
and fixed by 5831a59 :smile:

**Comment by nikashitsa** [comment]
Good :+1:  Need to replace delete method. Otherwise files will be in S3 forever even after message remove.
https://github.com/RocketChat/Rocket.Chat/blob/develop/server/methods/deleteMessage.coffee#L36

**Comment by sampaiodiego** [comment]
nice catch @nikashitsa .. I'll do it!

**Pull request merged**